kernel/functional.cc

```diff
@@ -0,0 +1,853 @@
+					}
+				} else if (chunk.is_port()) {
+					DriveChunkPort port_chunk = chunk.port();
+					if (port_chunk.is_whole()) {
+						if (driver_map.celltypes.cell_output(port_chunk.cell->type, port_chunk.port)) {
+							Node node = enqueue_cell(port_chunk.cell, port_chunk.port);
+							factory.update_pending(pending, node);
+						} else {
+							DriveSpec driver = driver_map(DriveSpec(port_chunk));
+							check_undriven(driver, RTLIL::unescape_id(port_chunk.cell->name) + " port " + RTLIL::unescape_id(port_chunk.port));
+							factory.update_pending(pending, enqueue(driver));
+						}
+					} else {
+						DriveChunkPort whole_port(port_chunk.cell, port_chunk.port, 0, GetSize(port_chunk.cell->connections().at(port_chunk.port)));
+						Node node = factory.slice(enqueue(whole_port), port_chunk.offset, port_chunk.width);
+						factory.update_pending(pending, node);
+					}
+				} else if (chunk.is_constant()) {
+					Node node = factory.constant(chunk.constant());
+					factory.suggest_name(node, "$const" + std::to_string(chunk.size()) + "b" + chunk.constant().as_string());
+					factory.update_pending(pending, node);
+				} else if (chunk.is_multiple()) {
+					log_error("Signal %s has multiple drivers. This is not supported by the functional backend. "
+						"If tristate drivers are used, call tristate -formal to avoid this error.\n", log_signal(chunk));
+				} else if (chunk.is_none()) {
+					undriven(log_signal(chunk));
+				} else {
+					log_error("unhandled drivespec: %s\n", log_signal(chunk));
+					log_abort();
+				}
+			} else {
+				log_abort();
+			}
+		}
+	}
+};
+
+IR IR::from_module(Module *module) {
+	IR ir;
+    auto factory = ir.factory();
+    FunctionalIRConstruction ctor(module, factory);
+    ctor.process_queue();
+    ir.topological_sort();
+    ir.forward_buf();
+    return ir;
+}
+
+void IR::topological_sort() {
+    Graph::SccAdaptor compute_graph_scc(_graph);
+    bool scc = false;
+    std::vector<int> perm;
+    TopoSortedSccs toposort(compute_graph_scc, [&](int *begin, int *end) {
+        perm.insert(perm.end(), begin, end);
+        if (end > begin + 1)
+        {
+            log_warning("Combinational loop:\n");
+            for (int *i = begin; i != end; ++i) {
+				Node node(_graph[*i]);
+                log("- %s = %s\n", RTLIL::unescape_id(node.name()).c_str(), node.to_string().c_str());
+			}
+            log("\n");
+            scc = true;
+        }
+    });
+	for(const auto &[name, state]: _states)
+		if(state.has_next_value())
+			toposort.process(state.next_value().id());
+	for(const auto &[name, output]: _outputs)
+		if(output.has_value())
+			toposort.process(output.value().id());
+	// any nodes untouched by this point are dead code and will be removed by permute
+    _graph.permute(perm);
+    if(scc) log_error("The design contains combinational loops. This is not supported by the functional backend. "
+		"Try `scc -select; simplemap; select -clear` to avoid this error.\n");
```

Comment:
This might be an outstanding TODO but `simplemap` will leave around gate cells which are unsupported by the functional backend as of yet

Reply:
oh yeah gate-level cells was on my TODO list but i didn't get around to it for the deliverable